src/main/java/software/amazon/smithy/lsp/SmithyLanguageServer.java

```diff
@@ -417,46 +413,72 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
         LOGGER.info("DidChangeWatchedFiles");
         // Smithy files were added or deleted to watched sources/imports (specified by smithy-build.json),
         // or the smithy-build.json itself was changed
-        Set<String> createdSmithyFiles = new HashSet<>(params.getChanges().size());
-        Set<String> deletedSmithyFiles = new HashSet<>(params.getChanges().size());
-        boolean changedBuildFiles = false;
-        for (FileEvent event : params.getChanges()) {
-            String changedUri = event.getUri();
-            if (changedUri.endsWith(".smithy")) {
-                if (event.getType().equals(FileChangeType.Created)) {
-                    createdSmithyFiles.add(changedUri);
-                } else if (event.getType().equals(FileChangeType.Deleted)) {
-                    deletedSmithyFiles.add(changedUri);
-                }
-            } else if (changedUri.endsWith(ProjectConfigLoader.SMITHY_BUILD)
-                    || changedUri.endsWith(ProjectConfigLoader.SMITHY_PROJECT)) {
-                changedBuildFiles = true;
-            } else {
-                for (String extFile : ProjectConfigLoader.SMITHY_BUILD_EXTS) {
-                    if (changedUri.endsWith(extFile)) {
-                        changedBuildFiles = true;
-                        break;
-                    }
-                }
+
+        Map<String, ProjectChanges> changesByProject = projects.computeProjectChanges(params.getChanges());
+
+        changesByProject.forEach((projectName, projectChanges) -> {
+            Project project = projects.getProjectByName(projectName);
+            if (projectChanges.hasChangedBuildFiles()) {
```

Comment:
I think we should short circuit this if else check for cases with no changes with `if (!projectChanges.hasChangedBuildFiles() || projectChanges.hasChangedSmithyFiles()) { return; }` unless the `ProjectChanges` can never be empty in that way?

Reply:
build files and smithy file changes aren't mutually exclusive, need to fix this conditional

Reply:
Actually, looking at this again, when build files are changed it will perform the work that would have been done for added/removed smithy files, so the conditional is correct